### PR TITLE
[JENKINS-36897] Disk Pool Restriction Extension Point

### PR DESCRIPTION
**Pull request by alexsomai**
Related to issue [[JENKINS-36897]](https://issues.jenkins-ci.org/browse/JENKINS-36897)

Summary of this pull request:
- [X] Implement the DiskPoolRestriction Extension Point
- [X] Write two implementations for the DiskPoolRestriction: 
 - [X] NoDiskPoolRestriction
 - [X] WhitelistUserRestriction
- [ ] Add javadocs and help files
- [ ] Write tests
- [ ] Write documentation

CC @martinda @oleg-nenashev

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -65,6 +66,13 @@
             <version>${workflow.version}</version>
         </dependency>
 
+        <!-- Temporary - needed for the UserSelector describable -->
+        <dependency>
+            <groupId>com.synopsys.arc.jenkinsci.plugins</groupId>
+            <artifactId>job-restrictions</artifactId>
```
Comment (oleg-nenashev) [inline pom.xml]: If you depend on it, maybe makes sense to implement some restriction strategies based on this plugin later. Now there is an issue with Pipeline compatibility, so it won't work correctly.

Copy-pasting of the code and restricting it maybe an option.

File: src/main/java/org/jenkinsci/plugins/ewm/restrictions/AbstractUserRestriction.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import javax.annotation.Nonnull;
+
+/**
+ * @author Alexandru Somai
+ */
+public abstract class AbstractUserRestriction extends DiskPoolRestriction {
+
+    @Override
+    public boolean isAllowed(@Nonnull Run<?, ?> run, @Nonnull TaskListener listener) {
+        Cause.UserIdCause userIdCause = run.getCause(Cause.UserIdCause.class);
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/ewm/restrictions/AbstractUserRestriction.java]: There is a corner-case, which may impact compatibility with other plugins. In old Hudson instances/plugins there was another cause - UserCause. It's deprecated for a long time, so you can just ignore it

File: src/main/java/org/jenkinsci/plugins/ewm/DiskPoolRestriction.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import hudson.model.AbstractDescribableImpl;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+
+import javax.annotation.Nonnull;
+
+/**
+ * @author Alexandru Somai
+ */
+public abstract class DiskPoolRestriction extends AbstractDescribableImpl<DiskPoolRestriction> implements ExtensionPoint {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/ewm/DiskPoolRestriction.java]: Good practice: add static all() method

**Comment by oleg-nenashev** [comment]
Looks good so far

File: src/main/java/org/jenkinsci/plugins/ewm/restrictions/AbstractUserRestriction.java (hunk 2 above)
Comment (martinda) [inline src/main/java/org/jenkinsci/plugins/ewm/restrictions/AbstractUserRestriction.java]: I do not understand why this class enforces the user to have triggered the build. What if the build is triggered by another build? Don't we have to look "deeper", i.e. for the initial cause of the upstream build?

File: src/main/resources/org/jenkinsci/plugins/ewm/restrictions/WhitelistUserRestriction/config.jelly [hunk 4]
```diff
@@ -0,0 +1,6 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
+    <f:entry title="${%Whitelisted users}">
+        <f:repeatableProperty field="usersList" add="${%Whitelist user}"/>
```
Comment (martinda) [inline src/main/resources/org/jenkinsci/plugins/ewm/restrictions/WhitelistUserRestriction/config.jelly]: Should the user list be made to look more like the user list [here](https://wiki.jenkins-ci.org/download/attachments/10059796/owndb.png?version=1&modificationDate=1202251054000)?

**Comment by alexsomai** [comment]
At our meeting we have discussed that there may be no need to provide an extension point for Disk Pool restriction. It can be easily done by integrating the Job Restriction plugin.
I'll be closing this in favor of #28.

**Pull request closed without merging**